### PR TITLE
[Bilibili] Add subtitle converter

### LINKED ISSUE #1015: [Site Request] Add subtitles conversion functionality for Bilibili's proprietary subtitle format
## Checklist

- [x] I'm reporting a site feature request
- [x] I've verified that I'm running yt-dlp version **2021.09.02** (I'm using the upstream version of yt-dlp, after the commit of Bilibili international)
- [x] I've searched the bug tracker for similar site feature requests including closed ones

## Description

The subtitle of Bilibili is delivered in BCC format (Bilibili proprietary subtitle format, `.json`). Currently, embedding this format of subtitles cannot be achieved because neither yt-dlp nor ffmpeg has the ability to do so.

The following terminal output can be used as an example. 
```
C:\Users\Tran Bich Dung>yt-dlp https://www.biliintl.com/en/play/37976/10344255 --write-subs --embed-subs --cookies "C:\Users\Tran Bich Dung\Downloads\fd798572-8600-4b99-b54a-ec690f6fc323.txt"
[BiliIntl] 10344255: Downloading JSON metadata
[BiliIntl] 10344255: Downloading JSON metadata
[BiliIntl] 10344255: Downloading JSON metadata
[info] 10344255: Downloading 1 format(s): 6+2
[info] Writing video subtitles to: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].en.json
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].en.json
[download] 100% of 23.50KiB in 00:00
[info] Writing video subtitles to: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].th.json
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].th.json
[download] 100% of 36.70KiB in 00:00
[info] Writing video subtitles to: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].vi.json
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].vi.json
[download] 100% of 25.60KiB in 00:00
[info] Writing video subtitles to: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].id.json
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].id.json
[download] 100% of 26.75KiB in 00:00
[SubtitlesConvertor] Converting subtitles
[SubtitlesConvertor] You have requested to convert json subtitles into another format, which is currently not possible
[SubtitlesConvertor] You have requested to convert json subtitles into another format, which is currently not possible
[SubtitlesConvertor] You have requested to convert json subtitles into another format, which is currently not possible
[SubtitlesConvertor] You have requested to convert json subtitles into another format, which is currently not possible
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].f6.mp4
[download] 100% of 131.17MiB in 03:00
[download] Destination: I'm Luffy! The Man Who Will Become the Pirate King! [10344255].f2.mp4
[download] 100% of 23.41MiB in 00:24
[Merger] Merging formats into "I'm Luffy! The Man Who Will Become the Pirate King! [10344255].mkv"
Deleting original file I'm Luffy! The Man Who Will Become the Pirate King! [10344255].f6.mp4 (pass -k to keep)
Deleting original file I'm Luffy! The Man Who Will Become the Pirate King! [10344255].f2.mp4 (pass -k to keep)
ERROR: 'duration'
```
To add the conversion functionality to yt-dlp, you can consult [this python script](https://github.com/y2361547758/bcc2ass). I want to thank @riyuejz for pointing this out.

Please note that the link in the output above is region-locked because that series is not available on Bilibili outside Asia. Thank you!

### PR DESCRIPTION
**Pull request by u-spec-png**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes https://github.com/yt-dlp/yt-dlp/issues/1015, code taken from https://github.com/y2361547758/bcc2ass.

### CODE REVIEW

**Comment by Ashish0804** [comment]
How can you be the original author as well as not the original author? 
Fix it please.

**Comment by pukkandan** [comment]
@Ashish0804 He wrote some of it, while other parts are from another repo

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/bilibili.py [hunk 1]
```diff
@@ -802,8 +827,8 @@ def _get_formats(self, type, ep_id):
         video_json = video_json['playurl']
         formats = []
         for vid in video_json.get('video', []):
-            video_res = vid.get('video_resource') or {}
-            video_info = vid.get('stream_info') or {}
+            video_res = vid.get('video_resource', {})
+            video_info = vid.get('stream_info', {})
```
[the remark this answers is not in the document]
Comment (u-spec-png) [reply yt_dlp/extractor/bilibili.py:831]: Isn't this the same as `or`?  [docs](https://docs.python.org/3/library/stdtypes.html#dict.get)

[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/bilibili.py:831]: Try running `{'key': None}.get('key', 'default')` and `{'key': None}.get('key') or 'default'` to see the difference

File: yt_dlp/extractor/bilibili.py [hunk 2]
```diff
@@ -783,15 +789,34 @@ class BiliIntlBaseIE(InfoExtractor):
...
+        return data
+
     def _get_subtitles(self, type, ep_id):
         sub_json = self._call_api(type, f'/m/subtitle?ep_id={ep_id}&platform=web', ep_id)
         subtitles = {}
         for sub in sub_json.get('subtitles', []):
             sub_url = sub.get('url')
             if not sub_url:
                 continue
+            sub_data = self._download_json(sub_url, ep_id)
```
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/bilibili.py]: you didnt add note

**Comment by pukkandan** [comment]
btw, is there any code remaining from https://github.com/y2361547758/bcc2ass now? It looks to me like you completely rewrote everything

**Comment by u-spec-png** [comment]
> btw, is there any code remaining from https://github.com/y2361547758/bcc2ass now? It looks to me like you completely rewrote everything

no

**Comment by riyuejz** [comment]
yt-dlp -F --cookies '/home/lynx/ytdlp/cookies.txt' https://www.biliintl.com/en/play/1005426/10499657
[BiliIntl] 10499657: Downloading JSON metadata
[BiliIntl] 10499657: Downloading JSON metadata
[BiliIntl] 10499657: Downloading JSON metadata
[BiliIntl] 10499657: Downloading JSON metadata
ERROR: _parse_json() missing 1 required positional argument: 'video_id'

**Pull request merged**